File: Snake/cpp/snake.cpp
#define OLC_PGE_APPLICATION
#include "olcPixelGameEngine.h"
#include <vector>
class Snake
{
public:
    std::vector<std::pair<int,int>> snake;
    int size;
    int dir;
    /*
        dir: 0 -> right
             1 -> down
             2 -> left
             3 -> up
    */
public:
    Snake(){
        snake.push_back(std::make_pair(0,-1));
        snake.push_back(std::make_pair(0,0));
        size = 2;
        dir = 0;
    }

    bool isColliding(){
        bool collided = false;
        for(int i=0;i<size-1;i++){
            if(snake[size-1].first==snake[i].first && snake[size-1].second==snake[i].second)
            {
                collided = true;
                break;
            }
        }
        return collided;
    }

    void move(bool grow){
        std::pair<int,int> new_head_pos = std::make_pair(snake[size-1].first,snake[size-1].second);
        if(dir==0)new_head_pos.second += 1;//right
        if(dir==1)new_head_pos.first += 1;//down
        if(dir==2)new_head_pos.second -= 1;//left
        if(dir==3)new_head_pos.first -= 1;//up
        if(grow){
            snake.push_back(new_head_pos);
            size++;
        }
        else{
            for(int i=0;i<size-1;i++){
                snake[i].first = snake[i+1].first;
                snake[i].second = snake[i+1].second;
            }
            snake[size-1].first = new_head_pos.first;
            snake[size-1].second = new_head_pos.second;
        }
    }

    void reset(){
        snake.clear();
        snake.push_back(std::make_pair(0,-1));
        snake.push_back(std::make_pair(0,0));
        size = 2;
        dir = 0;
    }
};

class Game : public olc::PixelGameEngine
{
private:
    Snake snake;
    float time;
    olc::HWButton left,right,up,down;
    std::pair<int,int> apple;

public:
	Game()
	{
		sAppName = "Snake";
	}

public:
    bool isColliding(){
        bool collided = false;
        if(fabs(snake.snake[snake.size-1].first)>ScreenHeight()/2)collided=true;
        if(fabs(snake.snake[snake.size-1].second)>ScreenWidth()/2)collided=true;
        collided |= snake.isColliding();
        return collided;
    }

    void setApple(){
        apple.first = rand()%(ScreenHeight()) - ScreenHeight()/2;
        apple.second = rand()%(ScreenWidth()) - ScreenWidth()/2;
    }

	bool OnUserCreate() override
	{
        time=0;
		return true;
	}

	bool OnUserUpdate(float fElapsedTime) override
	{
        left = GetKey(olc::LEFT);
        right = GetKey(olc::RIGHT);
        up = GetKey(olc::UP);
        down = GetKey(olc::DOWN);
        if(left.bPressed && snake.dir!=0)snake.dir=2;
        else if(right.bPressed && snake.dir!=2)snake.dir=0;
        else if(up.bPressed && snake.dir!=1)snake.dir=3;
        else if(down.bPressed && snake.dir!=3)snake.dir=1;

        int width=ScreenWidth()/2;
        int height = ScreenHeight()/2;
        FillRect(0,0,width*2,height*2,olc::BLACK);
		for (int i = 0; i < snake.size; i++)
			Draw(snake.snake[i].second + width,snake.snake[i].first+ height, olc::GREEN);
        Draw(apple.second + width,apple.first + height,olc::RED);
        time+=fElapsedTime;
        if(time>0.1){
            time=0;
            if(apple.first==snake.snake[snake.size-1].first && apple.second==snake.snake[snake.size-1].second)
            {
                snake.move(true);
                setApple();
            }
            else
            snake.move(false);
            if(isColliding()){
                snake.reset();
                setApple();
            }
        }
		return true;
	}

    
};


int main()
{
	Game main;
	if (main.Construct(50,50,10,10))
		main.Start();
	return 0;
}

File: Snake/js/snake.js
var c = document.getElementById("main");
c.width = 400;
c.height = 400;
var g = c.getContext("2d");
var snake = [{x:200,y:200},{x:200,y:210}];
var dir=3;//left up right down: 1 2 3 4
window.addEventListener('keydown',this.changeDir,false);
var food_x = Math.round((c.width-10)*Math.random()/10)*10;
var food_y = Math.round((c.height-10)*Math.random()/10)*10;
var gameon=true;
function myFunc(){
	setTimeout(
	function onTick(){
		g.fillStyle = "#000000";
		g.fillRect(0,0,c.width,c.height);
		move();
		drawSnake();
		if(gameon)
		myFunc();
	},100);
}

myFunc();

function drawSnake(){
if(gameon){
	var l = snake.length;
	var i=0;
	g.fillStyle = "white";
	for(i=0;i<l;i++){
		g.fillRect(snake[i].x,snake[i].y,10,10);
	}
	g.fillStyle = "red";
	g.fillRect(food_x,food_y,10,10);}
	else{
	g.fillStyle = "white";
	g.fillText("Game Over",180,190);
	g.fillText("Press any key to restart",180,210);
	}
}
function move(){
	var l = snake.length;
	var i=0;
	var tail = {x:snake[l-1].x,y:snake[l-1].y};
	for(i=l-1;i>0;i--){
		snake[i].x = snake[i-1].x;
		snake[i].y = snake[i-1].y;
	}
	if(dir==1)
		snake[0].x-=10;
	else if(dir==2)
		snake[0].y-=10;
	else if(dir==3)
		snake[0].x+=10;
	else if(dir==4)
		snake[0].y+=10;
	if(collided()){
		gameon=false;
	}
	if(snake[0].x<0)gameon=false;
	if(snake[0].y<0)gameon=false;
	if(snake[0].x>400)gameon=false;
	if(snake[0].y>400)gameon=false;
	if(snake[0].x == food_x && snake[0].y==food_y){
		snake[l] = tail;
		food_x = Math.round((c.width-10)*Math.random()/10)*10;
		food_y = Math.round((c.height-10)*Math.random()/10)*10;
	}
}

function collided(){
	var l = snake.length;
	var i=0;
	var j=0;
	for(i=0;i<l;i++){
		for(j=0;j<l;j++){
			if(i!=j && snake[i].x==snake[j].x && snake[i].y==snake[j].y)
			return true;
		}
	}
	return false;
}

function changeDir(event){
	if(gameon){
	var key = dir;
	if(event.keyCode==37 && dir!=3)dir=1;
	else if(event.keyCode==38 && dir!=4)dir=2;
	else if(event.keyCode==39 && dir!=1)dir=3;
	else if(event.keyCode==40 && dir!=2)dir=4;}
	else{gameon=true;myFunc();snake = [{x:200,y:200},{x:200,y:210}];dir=3;}
}


